File: 0160-intersection-of-two-linked-lists/striver_optimized_short_approach.cpp
// Time complexity: O(M + N), Space complexity: O(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // Cover the lists twice so that both the head pointers are directly aligned.
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* currA = headA;
        ListNode* currB = headB;

        while (currA != currB) {
            currA = currA == NULL ? headB : currA->next;
            currB = currB == NULL ? headA : currB->next;
        }

        return currA;
    }
};

File: 0160-intersection-of-two-linked-lists/striver_optimized_short_approach.py
# Time complexity: O(N + M), Space complexity: O(1)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        p1 = headA
        p2 = headB

        while p1 != p2:
            p1 = headB if p1 is None else p1.next
            p2 = headA if p2 is None else p2.next

        return p1